design-proposals/determining-host-model-migratability-level.md

```diff
@@ -0,0 +1,123 @@
+and nodes are being added\removed to\from the cluster.
+
+Even if a user has this information he would have to set nodeAffinity himself to prefer/require
+nodes only for the first time he schedule a vm.
+
+Therefore some users might benefit from such label and configuration options.
+
+## Goals
+- Add a label to each schedulable node in the cluster with a value of
+the relative amount of nodes that a vm with `host-model` cpuModel can
+migrate to when the node is the initial node of the vm.
+
+
+- Allow users to configure kubevirtCR to add node affinity for host-model 
+migratability level to virt-launcher when scheduling a vmi for the first time.
+
+## Non Goals
+Not changing the default form in which we schedule virtual machines.
+
+## Definition of Users
+Every user that is interested of migration of vms in heterogeneous cluster.
+
+## User Stories
+- As a user with a long living heterogeneous cluster, I'm **SUPRISED** to encounter host-model migration failures and
+i don't know how to prevent the failures.
+
+
+- As a user that use Kubevirt's tests in heterogeneous cluster i don't know why the tests are flacky:
+
+    In kubevirt-ci the cluster is homogeneous, so we aren't exposed to test failures that
+occur because of a host-model migration failures.
+
+  For instance Openshift-Virtualization project run our functional tests in heterogeneous
+clusters this issue cause flakiness and it is hard to follow.
+
+  
+Such configuration option that prefer/require host-model migratible nodes can solve this.
+
+## Repos
+https://github.com/kubevirt/kubevirt
+
+# Design
+- List and calculate the value of the relative amount of nodes that a vm with host-model 
+cpuModel can be scheduled to for each node  when adding\removing a node from the cluster.
+
+  This can be done through the `node-controller` which is 
+part of `virt-controller`. 
+
+
+- Add an optional fields to `kubevirt.spec.configuration.migrations` to determine if we want our vm's to prefer\require 
+nodes with at least `<some-numerical-number>` value of the new `kubevirt/host-model-migratability-level`
+label.
+
+## API Examples
+- add the following label to a node that allows host-model migration to 50% of the
+  cluster's schedulable nodes.
+```
+kubevirt/host-model-migratability-level: 50
+```
+
+and allow users to configure kubevirtCR to add the following node affinity to virt-launcher
+when scheduling a vmi for the first time.
+
+something like:
+
+```
+affinity:
+  nodeAffinity:
+    requiredDuringSchedulingIgnoredDuringExecution:
+      nodeSelectorTerms:
+        - matchExpressions:
+          - key: kubevirt/host-model-migratability-level
+            operator: Gt
+            values:
+            - <some-numerical-number>
+```
+
+
+- add fields to kubevirtCR:
+`kubevirt.spec.configuration.migrations.requireHostModelMigratabilityLevel:<some-number> `
+`kubevirt.spec.configuration.migrations.preferedHostModelMigratabilityLevel:<some-number> `
```

Comment:
Maybe this can be shortened to `requireMigratabilityLevel`?

I think that most of the users don't know about host-model (which is the default). This setting would not apply only for VMs that explicitly set their CPU model to a concrete model, and these are usually advanced users that know the implications.

Reply:
I think that this might mislead users to think that these configuration are valid to ANY CPU Model.